runtime/sema/check_variable_declaration.go

```diff
@@ -240,3 +242,169 @@ func (checker *Checker) elaborateNestedResourceMoveExpression(expression ast.Exp
+		case *ast.IdentifierExpression:
+			variableRefExpr = &refExpr.Identifier
+		default:
+			continue
+		}
+
+		if variableRefExpr == nil {
+			continue
+		}
+
+		referencedVariable := checker.valueActivations.Find(variableRefExpr.Identifier)
+		if referencedVariable == nil {
+			continue
+		}
+
+		// If the referenced variable is again a reference,
+		// then find the variable of the root of the reference chain.
+		// e.g.:
+		//     ref1 = &v
+		//     ref2 = &ref1
+		//     ref2.field = 3
+		//
+		// Here, `ref2` refers to `ref1`, which refers to `v`.
+		// So `ref2` is actually referring to `v`
+
+		referencedVars := nestedReferencedVariables(referencedVariable)
+		if referencedVars != nil {
+			variables = append(variables, referencedVars...)
+		}
+	}
+
+	return
+}
+
+// referenceExpressions returns all sub-expressions that may produce a reference.
+//
+// There could be two types of expressions that can result in a reference:
+//  1. Expressions that create a new reference.
+//     (i.e: reference-expression)
+//  2. Expressions that return an existing reference.
+//     (i.e: identifier-expression/member-expression/index-expression having a reference type)
+//
+// However, it is currently not possible to track member-expressions and index-expressions.
+// So this method either returns a reference-expression or an identifier-expression.
+//
+// The expression could also be hidden inside some other expression.
+// e.g(1): `&v as &T` is a casting expression, but has a hidden reference expression.
+// e.g(2): `(&v as &T?)!
+func referenceExpressions(expr ast.Expression) []ast.Expression {
+	switch expr := expr.(type) {
+	case *ast.ReferenceExpression:
+		return []ast.Expression{expr}
+	case *ast.ForceExpression:
+		return referenceExpressions(expr.Expression)
+	case *ast.CastingExpression:
+		return referenceExpressions(expr.Expression)
+	case *ast.BinaryExpression:
+		if expr.Operation != ast.OperationNilCoalesce {
+			return nil
+		}
+
+		refExpressions := make([]ast.Expression, 0)
+
+		lhsRef := referenceExpressions(expr.Left)
+		if lhsRef != nil {
+			refExpressions = append(refExpressions, lhsRef...)
+		}
+
+		rhsRef := referenceExpressions(expr.Right)
+		if rhsRef != nil {
+			refExpressions = append(refExpressions, rhsRef...)
+		}
+
+		return refExpressions
+	case *ast.ConditionalExpression:
+		refExpressions := make([]ast.Expression, 0)
+
+		thenRef := referenceExpressions(expr.Then)
+		if thenRef != nil {
+			refExpressions = append(refExpressions, thenRef...)
+		}
+
+		elseRef := referenceExpressions(expr.Else)
+		if elseRef != nil {
+			refExpressions = append(refExpressions, elseRef...)
+		}
+
+		return refExpressions
+	case *ast.IdentifierExpression:
+		return []ast.Expression{expr}
+	default:
+		return nil
+	}
+}
+
+// rootVariableOfExpression returns the identifier expression
+// of a var-ref/member-access/index-access expression.
+func rootVariableOfExpression(expr ast.Expression) *ast.Identifier {
```

Comment:
does go support tail recursion? this could be rewritten to a loop to avoid stack overflow

Reply:
refactored in 62b2e043a55cc0f15bf54632c2443494a04392ba